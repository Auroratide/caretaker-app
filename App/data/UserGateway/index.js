export class Permissions {
    static FOCUS = 'focus';
    static ADMIN = 'admin';
    static HELPER = 'helper';
}

const createUserGateway = function() {
    let allUsers = null;

    return {
        currentUser: null,

        getAll() {
            if(allUsers === null) {
                //api call goes here
                allUsers = [
                    {name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                    {name: 'Sarah', identifier: 's', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                    {name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone:'123-456-7890'},
                    {name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend']},
                    {name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone:'123-456-7890'}
                ]
            }
            return allUsers;
        },

        getName(identifier) {
            const user = allUsers.find((user) => user.identifier === identifier)
            return user ? user.name : 'TBD';
        },

        getRoles() {
            return this.currentUser.roles
        },

        signIn(email) {
            //make call to google oauth and store the results
            let results = {
                email: email
            };
            let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
            if(typeof userFromBackend !== 'undefined') {
                this.currentUser = userFromBackend;
                return true;
            }
            return false;
        },

        isSelf(identifier) {
            return this.currentUser.identifier === identifier;
        },

        isAdmin() {
            return this.currentUser.permission === Permissions.ADMIN
                || this.currentUser.permission === Permissions.FOCUS;
        },

        getCurrentPermissions() {
            return this.currentUser.permission;
        },

        inviteUser(email) {
            //api call to invite the user
            alert('invited ' + email + '!');
        }
    }
}

export default createUserGateway();
