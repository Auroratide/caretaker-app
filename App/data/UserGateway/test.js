import UserGateway from './index.js';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    })
});
