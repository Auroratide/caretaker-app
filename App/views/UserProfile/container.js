import React, { Component } from 'react';
import UserGateway, { NotificationTypes } from '../../data/UserGateway/';
import ScreenWithToolbar from '../../components/shared/ScreenWithToolbar';
import { createUserSettings, User } from '../../data/UserGateway';

export default ContainerFor = (UserProfile) => class extends Component {
    constructor() {
        super();
        const defaultUser = new User();
        this.state = {
            currentUser: defaultUser,
            settings: this.settingsFor(defaultUser)
        }
    }

    componentDidMount() {
        const user = UserGateway.getCurrentUser()
        this.setState({
            currentUser: user,
            settings: this.settingsFor(user)
        });
    }

    settingsFor(user) {
        let settings = createUserSettings();
        Object.keys(settings).forEach(key => {
            settings[key].value = user.notificationTypes.includes(key) || user.notificationTriggers.includes(key);
        })
        return settings;
    }

    toggle = (key) => () => {
        let settings = this.state.settings;
        const currentValue = settings[key].value
        settings[key].value = !currentValue;
        UserGateway.updateSettings(key, !currentValue);
        this.setState({
            currentUser: this.state.currentUser,
            settings: settings
        });
    }

    render() {
        const user = this.state.currentUser;
        return <ScreenWithToolbar title='Profile' navigation={this.props.navigation}>
            <UserProfile
                name={user.name}
                email={user.identifier}
                phone={user.phone}
                settings={this.state.settings}
                toggle={this.toggle}
            />
        </ScreenWithToolbar>
    }
}
