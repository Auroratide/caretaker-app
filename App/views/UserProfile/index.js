import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View
} from 'react-native';
import ContainerFor from './container.js';
import Checkbox from '../../components/material-ui/Checkbox.js';
import { NotificationTypes, NotificationTriggers } from '../../data/UserGateway';

const UserProfile = (props) => {

    _renderCheckbox = (type) => {
        const currentValue = props.notificationTypes[type].value;

        return <Checkbox
            label={props.notificationTypes[type].name}
            value={type}
            checked={currentValue}
            onCheck={props.toggle(type)}
        />
    }

    //todo remove this
    _renderNotificationTypes = () => {
        return Object.values(props.notificationTypes)
            .filter(type => type.value === true)
            .map(type => type.name)
            .join(', ')
    }

    return <View style={styles.container}>
        <Text style={styles.title}>User Profile</Text>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Info</Text>
            <Text>Name: {props.name}</Text>
            <Text>Email: {props.email}</Text>
            <Text>Phone: {props.phone}</Text>
        </View>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Settings</Text>
            <Text>Receive Notifications via: {_renderNotificationTypes()}</Text>
            <Text>Receive Notifications for: {props.notificationTriggers.join(', ')}</Text>
        </View>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Receive Notifications via:</Text>
            {_renderCheckbox(NotificationTypes.PUSH)}
            {_renderCheckbox(NotificationTypes.EMAIL)}
            {_renderCheckbox(NotificationTypes.SMS)}
        </View>
    </View>
};

export default ContainerFor(UserProfile);

const styles = StyleSheet.create({
    button: {
        margin: 10,
        padding: 10
    },
    container: {
        flex: 1,
        backgroundColor: '#F5FCFF',
        justifyContent: 'flex-start'
    },
    title: {
        fontSize: 24,
        textAlign: 'center',
        margin: 15,
    },
    sectionContainer: {
        margin: 10,
        flex: 1
    },
    sectionTitle: {
        fontSize: 16,
        fontWeight: 'bold',
        marginTop: 10,
    }
});
