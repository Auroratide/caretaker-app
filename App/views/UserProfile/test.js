import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import UserGateway, { NotificationTypes, NotificationTriggers } from '../../data/UserGateway/';
import UserProfile from './index.js';
import { Record } from 'immutable';

describe('User Profile', () => {

    beforeEach(() => {
        UserGateway.signIn('s');
    });

    it('snapshot', () => {
        expect(renderer.create(<UserProfile />)).toMatchSnapshot();
    })

    it('should process the raw user settings', () => {
        const profile = renderer.create(<UserProfile />).getInstance();
        expect(profile.state.settings).toEqual({
            [NotificationTypes.PUSH]: {
                name: 'Push',
                value: true
            },
            [NotificationTypes.EMAIL]: {
                name: 'Email',
                value: false
            },
            [NotificationTypes.SMS]: {
                name: 'SMS',
                value: false
            },
            [NotificationTriggers.ANNOUNCEMENT]: {
                name: 'Announcements',
                value: true
            },
            [NotificationTriggers.EVENT]: {
                name: 'Events',
                value: false
            },
            [NotificationTriggers.EVENT_BY_ROLE]: {
                name: 'Events specific to your role',
                value: true
            }
        });
    })

    it('should toggle the notification types', () => {
        const profile = renderer.create(<UserProfile />).getInstance();

        profile.toggle(NotificationTypes.EMAIL)();

        expect(profile.state.settings).toEqual({
            [NotificationTypes.PUSH]: {
                name: 'Push',
                value: true
            },
            [NotificationTypes.EMAIL]: {
                name: 'Email',
                value: true
            },
            [NotificationTypes.SMS]: {
                name: 'SMS',
                value: false
            },
            [NotificationTriggers.ANNOUNCEMENT]: {
                name: 'Announcements',
                value: true
            },
            [NotificationTriggers.EVENT]: {
                name: 'Events',
                value: false
            },
            [NotificationTriggers.EVENT_BY_ROLE]: {
                name: 'Events specific to your role',
                value: true
            }
        });
    })

    it('should update the user settings through the gateway', () => {
        const spy = jest.spyOn(UserGateway, 'updateSettings');
        const profile = renderer.create(<UserProfile />).getInstance();

        profile.toggle(NotificationTypes.EMAIL)();

        expect(spy).toHaveBeenCalledWith(NotificationTypes.EMAIL, true);
    })
})
