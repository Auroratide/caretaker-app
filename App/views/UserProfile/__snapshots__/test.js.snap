// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Profile snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ThemeProvider
    uiTheme={
      Object {
        "palette": Object {
          "primaryColor": "#FF7575",
          "secondaryColor": "#E66969",
        },
        "toolbar": Object {
          "container": Object {
            "height": 50,
          },
        },
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 24,
          "margin": 15,
          "textAlign": "center",
        }
      }
    >
      User Profile
    </Text>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Info
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Name: 
        Sarah
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Email: 
        s
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Phone: 
        123-456-7890
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Receive Notifications via:
      </Text>
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Receive Notifications for:
      </Text>
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "primaryColor": "#FF7575",
              "secondaryColor": "#E66969",
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
          }
        }
      />
    </View>
  </View>
</View>
`;
